specifications/secure-launch-specification.rst

```diff
@@ -0,0 +1,719 @@
+be measured.
+
+:pcr: PCR to store the measurement.
+:cfg: The address or value to measure.
+:size: The size to measure.
+:evt_info: Label to be recorded in TPM Event Log.
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_efi_cfg_entry {
+        u16 pcr;
+        u16 reserved;
+        u64 cfg; /* address or value */
+        u32 size;
+        char evt_info[TPM_EVENT_INFO_LENGTH];
+    } __packed;
+
+Appendix A: Measuring the DRTM Policy
+=====================================
+
+While the DRTM TPM event log is itself proof of the DRTM policy used by Secure
+Launch, there may be motivation for the policy itself to be incorporated into
+the measurement chain. While this section does not address the possible
+motivations or the validity of those motivations, a possible use of the policy
+measurement can be used as the value to cap the DRTM PCRs. Regardless of
+motivations, this appendix is to provide guidance on how the policy might be
+measured in a meaningful way.
+
+TPM Extend Operation
+--------------------
+
+For clarity, the extend operation, denoted here on out as E(), is an order
+preserving, recursive, mapping function. Consider any hash function, denoted as
+H(), the extend operation, is defined as:
+
+|    Given,
+|        0 = sizeof(H) bytes of 0
+|        Objs = [ Obj_0, ..., Obj_n ]
+|    Then,
+|        E(Objs) = {
+|            E_0 = H( 0 | H(Obj_0) )
+|            E_n = H( E_(n-1) | H(Obj_n) )
+|        }
+
+Measuring the Policy
+---------------------
+
+Measuring the policy is not as simple as hashing the block of memory containing
+the policy. This will not work as the policy may contain memory addresses that
+have the potential to change on the next launch of the system. As a result
+there is a potential for the next launch not to have the same memory addresses
+in the policy, which would in turn render a different measurement.
+
+To make a meaningful measurement of the policy, the measurement must capture
+what entities were to be measured and what order they were to be measured. To
+capture the first half, the measurement must contain enough of a DRTM policy
+entry to capture its uniqueness. Considering DRTM Policy Entry above, a
+combination of PCR, Entity Type, and Event Info yields a unique identity for
+the entry. To capture ordering of the policy events, the extend operation can
+be used to render a unique value that reflects the order of the policy events.
+
+Using this logic, the resulting operation to measure the policy would be as:
+
+|    Given,
+|        Entry_n = PCR_n | EntityType_n | EventInfo_n
+|        Policy = [ Entry_0, ..., Entry_n ]
+|    Then,
+|        M_policy = E(Policy)
+
+The result, `M_policy`, will be a hash of the policy that can then be extended
+into one, or more if using as a cap value, PCR(s).
+
+Appendix B: Intel TXT OS2MLE
+============================
+
+The Intel TXT specification provides a provision for the pre-launch environment to
```

Comment:
Does it make sense to provide a reference to the specification?

Reply:
added